### PR TITLE
Auth

### LINKED ISSUE #21: grid system for project page
@Aksh-Bansal-dev currently if I open the project page projects are not displayed in a grid but in a list type format with each row consisting of only one card.
Please fix this in your next PR.

### PR DESCRIPTION
**Pull request by Aksh-Bansal-dev**
Issue: #17

### CODE REVIEW

File: server/.env.example [hunk 1]
```diff
@@ -0,0 +1,6 @@
+DATABASE_URL=
+CLIENT_URL=
+JWT_REFRESH_SECRET=
+JWT_SECRET=
+GOOGLE_CLIENT_ID=
+GOOGLE_CLIENT_SECRET=
```
Comment (MananJethwani) [inline server/.env.example:6]: @Aksh-Bansal-dev add a URL to the page where users need to register for google client id and secret.

File: server/src/index.ts [hunk 2]
```diff
@@ -36,6 +40,7 @@ mongoose.connect(
 // ROUTES
```
Comment (MananJethwani) [inline server/src/index.ts:39]: @Aksh-Bansal-dev I want to propose one thing......we should create tables(collection) in the database if a developer/user is setting up the database for the first time, just check and create empty collections directly.
or am I forgetting something and it automatically creates one when it is referenced?

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  })
+);
+
+const App: React.FC = () => {
+  const classes = useStyles();
+  const [loading, setLoading] = React.useState(true);
+  React.useEffect(() => {
+    updateAccessToken(setLoading);
+  });
+  // if (loading) {
```
Comment (MananJethwani) [inline frontend/src/App.tsx]: should remove the comments or explain why they are present there.

**Review by MananJethwani — CHANGES_REQUESTED** [review]

File: frontend/src/App.tsx (hunk 3 above)
Comment (Aksh-Bansal-dev) [reply frontend/src/App.tsx]: Yeah, I forgot to remove unnecessary comments.

File: server/src/index.ts (hunk 2 above)
Comment (Aksh-Bansal-dev) [reply server/src/index.ts:39]: We just have to create a database in MongoDB, and the rest is handled by the mongoose. We don't need to create collections manually.

File: server/.env.example (hunk 1 above)
Comment (Aksh-Bansal-dev) [reply server/.env.example:6]: I think it will be better to add a URL to the official docs.

File: .husky/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+_
```
Comment (MananJethwani) [inline .husky/.gitignore:1]: is there actually a file named _ ?

File: frontend/src/components/RegistrationForm/index.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    },
+    or: {
+      margin: "2vh",
+    },
+  })
+);
+
+const RegistrationForm: React.FC = () => {
+  const classes = useStyles();
+  // const [login, setLogin] = useState(true);
```
Comment (MananJethwani) [inline frontend/src/components/RegistrationForm/index.tsx]: @Aksh-Bansal-dev more comments

File: frontend/src/components/RegistrationForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    <Paper className={classes.main}>
+      <Typography className={classes.title} align="center">
+        {/* {login ? "Login" : "Sign Up"} */}
+        {heading[page]}
+      </Typography>
+      <GoogleButton />
+      <Typography className={classes.or} align="center">
+        OR
+      </Typography>
+      {/* {login ? <Login setLogin={setLogin} /> : <Signup setLogin={setLogin} />} */}
```
Comment (MananJethwani) [inline frontend/src/components/RegistrationForm/index.tsx]: same here

File: frontend/src/pages/Projects/index.tsx [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import React from "react";
 import NavBar from "../../components/Navs/Navbar";
 import { createStyles, makeStyles } from "@material-ui/core/styles";
-import { Container } from "@material-ui/core";
+// import { Container } from "@material-ui/core";
```
Comment (MananJethwani) [inline frontend/src/pages/Projects/index.tsx]: remove these comment lines

File: frontend/src/pages/homePage/style.module.css [hunk 8]
```diff
@@ -1,10 +1,3 @@
-.fullpage {
-  /* display: grid;
```
Comment (MananJethwani) [inline frontend/src/pages/homePage/style.module.css:2]: same here

File: server/.env.example (hunk 1 above)
Comment (MananJethwani) [reply server/.env.example:6]: @Aksh-Bansal-dev no URL added here.

File: server/src/config/google-oauth.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import passport from "passport";
+import { Strategy as GoogleStrategy } from "passport-google-oauth20";
+import { User } from "../model/User";
+// eslint-disable-next-line @typescript-eslint/no-var-requires
```
Comment (MananJethwani) [inline server/src/config/google-oauth.ts:4]: same here

File: server/src/config/google-oauth.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    },
+    (_accessToken, _refreshToken, profile, done) => {
+      User.findOne({ googleId: profile.id }).then((currentUser) => {
+        if (currentUser) {
+          done(null, currentUser);
+        } else {
+          new User({
+            googleId: profile.id,
+            username: profile.displayName,
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (MananJethwani) [inline server/src/config/google-oauth.ts:22]: here as well

File: server/src/index.ts [hunk 11]
```diff
@@ -15,12 +18,13 @@ const app = express();
 app.use(express.json());
 app.use(
   cors({
-    // origin: process.env.CLIENT_URL,
-    origin: "*",
+    origin: process.env.CLIENT_URL,
+    // origin: "*",
```
Comment (MananJethwani) [inline server/src/index.ts]: same here

File: server/src/model/User.ts [hunk 12]
```diff
@@ -1,41 +1,54 @@
...
   username: string;
+  googleId?: string;
   email: string;
-  password: string;
+  password?: string;
   tokenVersion: number;
-  clubs?: string[];
-  authority: "member" | "admin";
+  clubs?: clubObj[];
+  // authority: "member" | "admin";
```
Comment (MananJethwani) [inline server/src/model/User.ts]: same here

File: server/src/model/User.ts [hunk 13]
```diff
@@ -1,41 +1,54 @@
...
+    {
+      name: {
+        type: String,
+      },
+      authority: {
+        type: String,
+      },
+    },
+  ],
+  // authority: {
```
Comment (MananJethwani) [inline server/src/model/User.ts]: same here

File: server/src/routes/googleAuthRoutes.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import express from "express";
+import passport from "passport";
+const Router = express.Router();
+
+Router.get(
+  "/",
+  passport.authenticate("google", {
+    scope: ["profile", "email"],
+    session: false,
+    // scope: ["https://www.googleapis.com/auth/plus.login"],
```
Comment (MananJethwani) [inline server/src/routes/googleAuthRoutes.ts]: here as well

File: server/src/routes/refreshtoken.ts [hunk 15]
```diff
@@ -4,44 +4,39 @@ import { getAccessToken, sendRefreshToken } from "../utils/tokenstuff";
 import { verify } from "jsonwebtoken";
 const Router = express.Router();
 
-Router.post("/refresh_token", async (req, res) => {
-  // console.log(req);
```
Comment (MananJethwani) [inline server/src/routes/refreshtoken.ts:8]: here

File: server/src/utils/tokenstuff.ts [hunk 16]
```diff
@@ -24,6 +24,7 @@ export const sendRefreshToken = (res: Response, user?: User): void => {
     ),
     {
       httpOnly: true,
+      // path: "/",
```
Comment (MananJethwani) [inline server/src/utils/tokenstuff.ts]: here

**Review by MananJethwani — APPROVED** [review]
just remove these comments and squash so that I can merge this PR thx

File: .husky/.gitignore (hunk 4 above)
Comment (Aksh-Bansal-dev) [reply .husky/.gitignore:1]: yes, it's for husky

File: server/src/config/google-oauth.ts (hunk 10 above)
Comment (Aksh-Bansal-dev) [reply server/src/config/google-oauth.ts:22]: this one is required to disable eslint for 1 line

**Pull request merged**